### PR TITLE
feat: try to improve object storage io read

### LINKED ISSUE #9308: performance: add pre-fetch for block reader
**Summary**

For sync read:
https://github.com/datafuselabs/databend/blob/523e2190c275d481c707f8ee35972fc7290cba38/src/query/storages/fuse/fuse/src/io/read/block_reader.rs#L380-L390

https://github.com/datafuselabs/databend/blob/523e2190c275d481c707f8ee35972fc7290cba38/src/query/storages/fuse/fuse/src/io/read/block_reader.rs#L389-L389

If we have many indices may be a waste because their offset may be adjacent (note: not sequentially connected, there may be some gaps, like the gap <1KB), converts small fragmented reads into one large read,  so we can merge the reading them all in one:
```
let (merge_read_offset, merge_read_length) = ...;
let result = Self::sync_read_column(op.object(&location), merge_read_offset, merge_read_length); 
-- Strip out (offset, length) data from result
```

### PR DESCRIPTION
**Pull request by BohuTANG**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

### Try to merge neighboring io operations using RangeMerger
* If the distance between two IO ranges to be read in one file is less than `storage_io_min_bytes_for_seek`, then Databend sequentially reads a range of files that contains both ranges, thus avoiding extra seek.
* This only works for the remote read.

### Introduce two remote read IO settings
* `storage_io_min_bytes_for_seek`: If the distance between two IO requests to be read in one file is less than storage_io_min_bytes_for_seek, merge them into one, the default value is 512Bytes.
* `storage_io_max_page_bytes_for_read`: The maximum bytes of one IO request to read. Default the value is 512KB

### Introduce remote IO request perf counter
* remote_io_seeks: the raw range numbers before IO merged
* remote_io_seeks_after_merged: the ranges need to read after IO merged
* remote_io_read_bytes: the bytes read before IO merged
* remote_io_read_bytes_after_merged:  the bytes need to read after IO merged
* remote_io_read_milliseconds: read cost in milliseconds

Case1:
```
SELECT * FROM hits WHERE URL LIKE '%google%' ORDER BY EventTime LIMIT 10 ignore_result;

-- No merged IO (set storage_io_min_bytes_for_seek=0;)
mysql> select metric, value  from system.metrics where metric like '%remote_io%' order by metric;
+----------------------------------------+---------------+
| metric                                 | value         |
+----------------------------------------+---------------+
| fuse_remote_io_read_bytes              | 15205646411.0 |
| fuse_remote_io_read_bytes_after_merged | 15205646411.0 |
| fuse_remote_io_read_milliseconds       | 546271.0      |
| fuse_remote_io_read_parts              | 1490.0        |
| fuse_remote_io_seeks                   | 77607.0       |
| fuse_remote_io_seeks_after_merged      | 77607.0       |
+----------------------------------------+---------------+

-- Merged IO (set storage_io_min_bytes_for_seek=48)
mysql> select metric, value  from system.metrics where metric like '%remote_io%' order by metric;
+----------------------------------------+---------------+
| metric                                 | value         |
+----------------------------------------+---------------+
| fuse_remote_io_read_bytes              | 15205646411.0 |
| fuse_remote_io_read_bytes_after_merged | 15208201850.0 |
| fuse_remote_io_read_milliseconds       | 446209.0      |
| fuse_remote_io_read_parts              | 1490.0        |
| fuse_remote_io_seeks                   | 77607.0       |
| fuse_remote_io_seeks_after_merged      | 13904.0       |
+----------------------------------------+---------------+
```
For this query, the total IO time was from 546271.0ms to 446209.0ms(after Merged IO).

Case2:
```
mysql> select * from hits ignore_result;
Empty set (19.74 sec)
Read 99997497 rows, 75.78 GiB in 19.733 sec., 5.07 million rows/sec., 3.84 GiB/sec.

mysql> select metric, value  from system.metrics where metric like '%remote_io%' order by metric;
+----------------------------------------+---------------+
| metric                                 | value         |
+----------------------------------------+---------------+
| fuse_remote_io_read_bytes              | 15412179252.0 |
| fuse_remote_io_read_bytes_after_merged | 15412179252.0 |
| fuse_remote_io_read_milliseconds       | 375970.0      |
| fuse_remote_io_read_parts              | 751.0         |
| fuse_remote_io_seeks                   | 78855.0       |
| fuse_remote_io_seeks_after_merged      | 78855.0       |
+----------------------------------------+---------------+


mysql> select * from hits ignore_result;
Empty set (19.71 sec)
Read 99997497 rows, 75.78 GiB in 19.701 sec., 5.08 million rows/sec., 3.85 GiB/sec.

mysql> select metric, value  from system.metrics where metric like '%remote_io%' order by metric;
+----------------------------------------+---------------+
| metric                                 | value         |
+----------------------------------------+---------------+
| fuse_remote_io_read_bytes              | 15412179252.0 |
| fuse_remote_io_read_bytes_after_merged | 15414801174.0 |
| fuse_remote_io_read_milliseconds       | 337415.0      |
| fuse_remote_io_read_parts              | 751.0         |
| fuse_remote_io_seeks                   | 78855.0       |
| fuse_remote_io_seeks_after_merged      | 13461.0       |
+----------------------------------------+---------------+
```

### Bench

Nightly-162 vs. This PR

image

Closes #9308

### CODE REVIEW

**Comment by BohuTANG** [comment]
Merge io request is always faster than normal:
image

**Comment by BohuTANG** [comment]
## Bench

16C32G

Raw IO: Nightly-159
Merged IO: this patch

image

**Comment by BohuTANG** [comment]
@mergify update

**Comment by BohuTANG** [comment]
@mergify update

**Comment by BohuTANG** [comment]
@mergify update

File: src/query/storages/fuse/src/io/read/block_reader.rs [hunk 1]
```diff
@@ -27,3 +41,119 @@ pub struct BlockReader {
...
+
+impl BlockReader {
+    /// This is an optimized for data read, works like the Linux kernel io-scheduler IO merging.
+    /// If the distance between two IO request ranges to be read is less than storage_io_min_bytes_for_seek(Default is 48Bytes),
+    /// will read the range that contains both ranges, thus avoiding extra seek.
+    ///
+    /// It will *NOT* merge two requests:
+    /// if the last io request size is larger than storage_io_page_bytes_for_read(Default is 512KB).
+    pub async fn merge_io_read(
+        ctx: Arc<dyn TableContext>,
```
Comment (zhang2014) [inline src/query/storages/fuse/src/io/read/block_reader.rs:53]: extract to read_settings is better?

Comment (BohuTANG) [reply src/query/storages/fuse/src/io/read/block_reader.rs:53]: Like this?
```
   pub async fn merge_io_read(
        object: Object,
        max_page_size:u64,
        min_seek_size:u64,
        raw_ranges: Vec<(usize, Range<u64>)>,
    ) -> Result<Vec<(usize, Vec<u8>)>> {
```

Comment (zhang2014) [reply src/query/storages/fuse/src/io/read/block_reader.rs:53]: yep

Comment (BohuTANG) [reply src/query/storages/fuse/src/io/read/block_reader.rs]: Done

**Review by zhang2014 — APPROVED** [review]

**Pull request merged**